### DISCUSSION TITLE
How to handle input file upload Cancel button when a file is already uploaded

### DISCUSSION THREAD

**Discussion by vinamelody** [react-hook-form/react-hook-form#2855] [Q&A]
Hi, I have an input file upload. Currently stuck with this behavior: a file is already uploaded, the I click Select File again and click Cancel.
- The validation triggers
- The validation fails

Desired behavior:
- Validation triggers but should pass

![Screenshot 2020-09-09 at 9 24 42 AM](https://user-images.githubusercontent.com/990511/92543756-e5988300-f27e-11ea-8467-6fd875d953c9.png)

Here's the complete CSB for the implementation <https://codesandbox.io/s/stupefied-borg-zvl16?file=/src/pages/Step2/index.js>

```javascript
// InputFile.js

const handleOnChange = e => {
    const files = e.target.files

    if (files.length > 0) {
      setFilename(e.target.files[0])
      uploadFileHandler(e)
    } else {
      // Possibly cancel button is clicked here
      if (filename) {
           // An uploaded file exist, how to tell form that this is valid?
           ?
      }
    }
  }
```

```javascript
// File handler in parent component
const handleUploadFile = useCallback(
  ({
    target: {
      validity,
      files: [file]
    }
  }) => {
    validity.valid && setUploadedFile(file)
  }, [])
```

**Comment by bluebill1049** [comment]
the codesandbox is rather large.  I would recommend reproducing the behavior in a simple and isolated CSB, not with the entire project. on top of that, i would recommend use `validate` function from register

**Reply by vinamelody** [reply]
Sorry, i removed all other form elements from the CSB. 

If I'm using yup schema for validation, can I also use `validation` from the doc page?
```
register({
      validate: value => value === '1'
    })
```

**Reply by bluebill1049** [reply]
oh you can't, if you are using schema, you have to stick with it.

**Reply by bluebill1049** [reply]
what i would do is to test out your form data? see what's in there: `getValues()` and then test again your schema, and see why the validation is poping up.

**Reply by vinamelody** [reply]
I wish to understand more about the file input and if i'm handling it correctly.
Even though the initial value is set to `''` , its actual value is handled by `onChange` function below.
Is there any way to keep it inside the `ref`?

```
// File handler in parent component
const handleUploadFile = useCallback(
  ({
    target: {
      validity,
      files: [file]
    }
  }) => {
    validity.valid && setUploadedFile(file)
  }, [])
```

**Reply by vinamelody** [reply]
The reason i'm asking this is because there's back & next movement for 5 steps form :( ... quite challenging..

i'm storing the File object in the reducer and need to set the file upload value again when user visit the page, hence user doesn't need to upload file but still need to validate again

**Reply by vinamelody** [reply]
> what i would do is to test out your form data? see what's in there: `getValues()` and then test again your schema, and see why the validation is poping up.

ah ok thanks for the pointer!

when i give up with the uncontrolled, is it a good idea to use Controlled? haha

**Reply by bluebill1049** [reply]
> The reason i'm asking this is because there's back & next movement for 5 steps form :( ... quite challenging..
> 
> i'm storing the File object in the reducer and need to set the file upload value again when user visit the page, hence user doesn't need to upload file but still need to validate again

hmm store the file object is probably a scary thing to do, consider file object is immutable. i don't think you can even do that, it's a security thing.

**Reply by vinamelody** [reply]
It's temporary until this cancel thing is resolved. 
Is there an example code to setValue for uncontrolled input file?

Can it be a File object rather than FileList?

**Reply by vinamelody** [reply]
Oops sorry, again I confused myself 😬 setting input value is by default not allowed + there's no way to detect Cancel event (which in my case, triggers validation)

**Reply by bluebill1049** [reply]
> It's temporary until this cancel thing is resolved.
> Is there an example code to setValue for uncontrolled input file?
> 
> Can it be a File object rather than FileList?

I am not sure you can store the input file object, I could be wrong, google it.
